Public/Src/App/Bxl/Args.cs

```diff
@@ -1205,47 +1213,32 @@ public bool TryParse(string[] args, PathTable pathTable, out ICommandLineConfigu
-                        }
-                        else
-                        {
-                            // Since i < limit, this is guaranteed to make progress.
-                            limit = i;
-                        }
-                    }
-
-                    // Check for equality
-                    if (min < m_handlers.Length)
-                    {
-                        if (string.Equals(m_handlers[min].OptionName + m_handlers[min].Suffix, opt.Name, StringComparison.OrdinalIgnoreCase))
-                        {
-                            // min it is
-                            m_handlers[min].Action(opt);
+                IterateArgs(cl, configuration, specialCaseUnsafeOptions);
 
-                            // compile list of user-enabled unsafe options to log later
-                            if (specialCaseUnsafeOptions.Contains(opt.Name) || (m_handlers[min].IsUnsafe && m_handlers[min].IsEnabled()))
-                            {
-                                configuration.CommandLineEnabledUnsafeOptions.Add(m_handlers[min].OptionName);
-                            }
-
-                            continue;
-                        }
-                    }
-
-                    // unknown argument, fail
-                    throw CommandLineUtilities.Error(Strings.Args_Args_NotRecognized, opt.Name);
+                int numABTestingOptions = loggingConfiguration.ABTestingArgs.Count;
+                if (numABTestingOptions > 0)
+                {
+                    // If RelatedActivityId is populated, use it as a seed for random number generation
+                    // so that we can use the same abTesting args for master-workers and different build phases (enlist, meta, product).
+                    Random randomGen = string.IsNullOrEmpty(loggingConfiguration.RelatedActivityId) ?
+                        new Random() :
+                        new Random(loggingConfiguration.RelatedActivityId.GetHashCode());
+
+                    int randomNum = randomGen.Next(numABTestingOptions);
+                    var randomOption = loggingConfiguration.ABTestingArgs.ToList()[randomNum];
+                    string abTestingId = randomOption.Key;
+                    string abTestingArgs = randomOption.Value;
+
+                    // Add key and the hash code of the arguments to traceInfo for telemetry purposes. 
+                    // As the ID for the AB testing arguments is specified by the user, there is a chance to
+                    // give the same ID to the different set of arguments. That's why, we also add the hash code of 
+                    // the arguments to traceInfo.
+                    loggingConfiguration.TraceInfo.Add(
+                        TraceInfoExtensions.ABTesting, 
+                        $"{abTestingId};{abTestingArgs.GetHashCode().ToString()}");
+
+                    var cl2 = new CommandLineUtilities(abTestingArgs.Split(' '));

```

Comment:
Is `abTestingArgs.Split(' ')` always going to be correct?  I don't think it will, since each BuildXL arg inside of `abTestingArgs` may contain spaces (e.g., `/abTesting:AB1="/f:spec='*/a b/spec.dsc'"` or something)

Reply:
This should do the right thing:
```csharp
BuildXL.Utilities.CLI.WinParser().SplitArgs(abTestingArgs);
```

Reply:
thanks, replaced with WinParser.